add some unit tests for estimateGas

yihuang opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

- add unit tests for estimateGas
- add some test environment setup, Closes #323
- add a precompiled erc20 smart contract in the test case, instruction to compile is included in the comment

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

yihuang commented:
One of the test cases currently fail for estimated gas number different from the one returned by geth.

fedekunze commented on x/evm/keeper/grpc_query.go:
```diff
@@ -406,6 +406,10 @@ func (k Keeper) EstimateGas(c context.Context, req *types.EthCallRequest) (*type
 	k.WithContext(ctx)
 
+	if req.GasCap < ethparams.TxGas {
+		return nil, status.Error(codes.InvalidArgument, "GasCap must not lower than 21000")
```
```suggestion
		return nil, status.Error(codes.InvalidArgument, "gas cap cannot be lower than 21,000")
```

fedekunze commented on x/evm/keeper/grpc_query_test.go:
```diff
@@ -18,8 +23,27 @@ import (
+	invalidAddress = "0x0000"
+
+	// Build with "solc --bin --abi --allow-paths . ./tests/solidity/suites/staking/contracts/test/mocks/StandardTokenMock.sol"
+	contractBinStr = "608060405234801561001057600080fd5b506040516113913803806113918339818101604052604081101561003357600080fd5b810190808051906020019092919080519060200190929190505050806000808473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff168152602001908152602001600020819055508060028190555050506112e8806100a96000396000f3fe608060405234801561001057600080fd5b50600436 ...
```
nit, can you move this to another file, eg: `keeper/abi_test.go`, when I open this file it crashes VSCode

yihuang replied:
sure, maybe both bin and abi, put together into a json file.

yihuang replied:
done, use `go:embed`, thus increase golang version to 1.16.

fedekunze reviewed (approved):
ACK, one minor suggestion

yihuang commented:
> One of the test cases currently fail for estimated gas number different from the one returned by geth.

the test case failure is due to unclean context, now added capability for test suite to commit and start a new block.

fedekunze commented:
> the test case failure is due to unclean context, now added capability for test suite to commit and begin a new block.

can you fix the failing test? Then we can merge this PR

yihuang commented:
> > the test case failure is due to unclean context, now added capability for test suite to commit and begin a new block.
> 
> can you fix the failing test? Then we can merge this PR

that one is already fixed, current failure seems caused by something else, I'm checking it now.

yihuang commented:
> > the test case failure is due to unclean context, now added capability for test suite to commit and begin a new block.
> 
> can you fix the failing test? Then we can merge this PR

fixed now, remove the commit in test suite setup.
